homeassistant/components/opnsense/__init__.py

```diff
@@ -0,0 +1,76 @@
+"""Support for OPNSense Routers."""
+import logging
+
+from pyopnsense import diagnostics
+import voluptuous as vol
+
+import homeassistant.helpers.config_validation as cv
+from homeassistant.const import CONF_URL, CONF_API_KEY, CONF_VERIFY_SSL
+from homeassistant.helpers.discovery import load_platform
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_API_SECRET = "api_secret"
+CONF_TRACKER_INTERFACE = "tracker_interfaces"
+
+DOMAIN = "opnsense"
+
+OPNSENSE_DATA = DOMAIN
+
+CONFIG_SCHEMA = vol.Schema(
+    {
+        DOMAIN: vol.Schema(
+            {
+                vol.Required(CONF_URL): cv.url,
+                vol.Required(CONF_API_KEY): cv.string,
+                vol.Required(CONF_API_SECRET): cv.string,
+                vol.Optional(CONF_VERIFY_SSL, default=False): cv.boolean,
+                vol.Optional(CONF_TRACKER_INTERFACE, default=[]): vol.All(
+                    cv.ensure_list, [cv.string]
+                ),
+            }
+        )
+    },
+    extra=vol.ALLOW_EXTRA,
+)
+
+
+def setup(hass, config):
+    """Set up the opnsense component."""
+
+    conf = config[DOMAIN]
+    url = conf[CONF_URL]
+    api_key = conf[CONF_API_KEY]
+    api_secret = conf[CONF_API_SECRET]
+    verify_ssl = conf[CONF_VERIFY_SSL]
+    tracker_interfaces = conf[CONF_TRACKER_INTERFACE]
+
+    interfaces_client = diagnostics.InterfaceClient(
+        api_key, api_secret, url, verify_ssl
+    )
+    try:
+        interfaces_client.get_arp()
+    except Exception:  # pylint: disable=broad-except
```

Comment:
Please catch a specific exception raised by the library on credentials or connection failure.

It's good if the library raises different exceptions for different type of problems.

Reply:
It depends on the failure mode, the library will raise a `pyopnsense.exceptions.APIException` if the api returns an error, because of invalid credentials for example. But if there is a connection issue or something else that happens at a lower level like requests or urllib3 that will get raised verbatim. Which is why I used a broad catch here. It seemed like it would be better to catch any exception and log it rather than raise an unhandled exception from a lower layer.

Reply:
Preferably the library should catch RequestException and raise a library specific error.